### PR TITLE
IGNITE-7149: Gradient Boosting on Trees

### PR DESCRIPTION
**Pull request by avplatonov**
Implementation GDB on DecisionTrees

### CODE REVIEW

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBTrainer.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    /**
+     * Estimate gradient value for answer-prediction pair.
+     *
+     * @param sampleSize Sample size.
+     * @param answer Right answer.
+     * @param prediction Prediction.
+     */
+    protected abstract double grad(long sampleSize, double answer, double prediction);
```
Comment (artemmalykh) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBTrainer.java]: I suggest use here composition instead of inheritance. Let's feed grad as lambda as one more constructor parameter. This way we'll be able to make for example more easy grid search on losses.

Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBTrainer.java]: ok.

File: examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GDBOnTreesClassificationTrainerExample.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import org.apache.ignite.thread.IgniteThread;
+
+/**
+ * Example represents a solution for the task of classification learning based on
+ * Gradient Boosting on trees implementation. It shows an initialization of {@link org.apache.ignite.ml.tree.boosting.GDBBinaryClassifierOnTreesTrainer},
+ * initialization of Ignite Cache, learning step and comparing of predicted and real values.
+ *
+ * In this example dataset is creating automatically by meander function f(x) = [sin(x) > 0].
+ */
+public class GDBOnTreesClassificationTrainerExample {
```
Comment (zaleslaw) [inline examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GDBOnTreesClassificationTrainerExample.java]: Maybe rename this class?

File: examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GDBOnTreesClassificationTrainerExample.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+     */
+    public static void main(String... args) throws InterruptedException {
+        // Start ignite grid.
+        try (Ignite ignite = Ignition.start("examples/config/example-ignite.xml")) {
+            System.out.println(">>> Ignite grid started.");
+
+            IgniteThread igniteThread = new IgniteThread(ignite.configuration().getIgniteInstanceName(),
+                GDBBinaryClassifierOnTreesTrainer.class.getSimpleName(), () -> {
+
+                // Create cache with training data.
```
Comment (zaleslaw) [inline examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GDBOnTreesClassificationTrainerExample.java]: Move cache configuration / data filling to separate methods

File: examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GDBOnTreesClassificationTrainerExample.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                    System.out.printf(">>> | %.4f\t\t| %.4f\t\t|\n", predicted, Math.sin(x) < 0 ? 0.0 : 1.0);
+                }
+
+                System.out.println(">>> ---------------------------------");
+
+                System.out.println(">>> GDB classification trainer example completed.");
+            });
+
+            igniteThread.start();
+
```
Comment (zaleslaw) [inline examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GDBOnTreesClassificationTrainerExample.java]: drop the line

File: examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GRBOnTreesRegressionTrainerExample.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * Example represents a solution for the task of regression learning based on
+ * Gradient Boosting on trees implementation. It shows an initialization of {@link org.apache.ignite.ml.tree.boosting.GDBRegressionOnTreesTrainer},
+ * initialization of Ignite Cache, learning step and comparing of predicted and real values.
+ *
+ * In this example dataset is creating automatically by parabolic function f(x) = x^2.
+ */
+public class GRBOnTreesRegressionTrainerExample {
+    /**
+     * Executes example.
```
Comment (zaleslaw) [inline examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GRBOnTreesRegressionTrainerExample.java]: Correct english

File: examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GRBOnTreesRegressionTrainerExample.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                // Create cache with training data.
+                CacheConfiguration<Integer, double[]> trainingSetCfg = new CacheConfiguration<>();
+                trainingSetCfg.setName("TRAINING_SET");
+                trainingSetCfg.setAffinity(new RendezvousAffinityFunction(false, 10));
+
+                IgniteCache<Integer, double[]> trainingSet = ignite.createCache(trainingSetCfg);
+                for(int i = -50; i <= 50; i++) {
+                    double x = ((double)i) / 10.0;
+                    double y = Math.pow(x, 2);
+                    trainingSet.put(i, new double[] {x, y});
```
Comment (zaleslaw) [inline examples/src/main/java/org/apache/ignite/examples/ml/tree/boosting/GRBOnTreesRegressionTrainerExample.java]: refactor ten previous lines to separate methods

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/ModelOnFeaturesSubspace.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Collections;
+import java.util.Map;
+import org.apache.ignite.ml.Model;
+import org.apache.ignite.ml.math.Vector;
+import org.apache.ignite.ml.math.VectorUtils;
+
+/**
+ * Model trained on a features subspace with mapping from original features space to subspace.
+ */
+public class ModelOnFeaturesSubspace implements Model<Vector, Double> {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/ModelOnFeaturesSubspace.java]: Maybe rename this class?

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBBinaryClassifierTrainer.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+import org.apache.ignite.ml.structures.partition.LabeledDatasetPartitionDataBuilderOnHeap;
+
+/**
+ * Trainer for binary classifier using Gradient Boosting.
+ * As preparing stage this algorithm learn labels in dataset and create mapping dataset labels to 0 and 1.
+ * This algorithm uses gradient of Logarithmic Loss metric [LogLoss] by default in each step of learning.
+ */
+public abstract class GDBBinaryClassifierTrainer extends GDBTrainer {
+    /** External representation of first class. */
+    private double externalFirstCls; //internal 0.0
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBBinaryClassifierTrainer.java]: I don't understand this comment

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBBinaryClassifierTrainer.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * Constructs instance of GDBBinaryClassifierTrainer.
+     *
+     * @param gradStepSize Grad step size.
+     * @param cntOfIterations Count of learning iterations.
+     * @param lossGradient Gradient of loss function. First argument is sample size, second argument is valid answer, third argument is current model prediction.
+     */
+    public GDBBinaryClassifierTrainer(double gradStepSize,
+        Integer cntOfIterations,
+        IgniteTriFunction<Long, Double, Double, Double> lossGradient) {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBBinaryClassifierTrainer.java]: Unreadable function)))

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBBinaryClassifierTrainer.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+                        if (a == null)
+                            return b;
+                        if (b == null)
+                            return a;
+                        a.addAll(b);
+                        return a;
+                    }
+                ));
+
+        A.ensure(uniqLabels.size() == 2, "Binary classifier expects two types of labels in learning dataset");
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBBinaryClassifierTrainer.java]: Unreadable code. Please rename or provide comment

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBTrainer.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    /** {@inheritDoc} */
+    @Override public <K, V> Model<Vector, Double> fit(DatasetBuilder<K, V> datasetBuilder,
+        IgniteBiFunction<K, V, double[]> featureExtractor,
+        IgniteBiFunction<K, V, Double> lbExtractor) {
+
+        learnLabels(datasetBuilder, featureExtractor, lbExtractor);
+
+        IgniteBiTuple<Double, Long> initAndSampleSize = computeInitialValue(datasetBuilder,
+            featureExtractor, lbExtractor);
+        Double mean = initAndSampleSize.get1();
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBTrainer.java]: avoid get1 and get2, it's not a Scala, please use meaninful names

File: modules/ml/src/main/java/org/apache/ignite/ml/knn/classification/KNNClassificationModel.java [hunk 12]
```diff
@@ -43,7 +43,7 @@
 /**
  * kNN algorithm model to solve multi-class classification task.
  */
-public class KNNClassificationModel<K, V> implements Model<Vector, Double>, Exportable<KNNModelFormat> {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/knn/classification/KNNClassificationModel.java]: What's the reason to drop these generic?

File: modules/ml/src/main/java/org/apache/ignite/ml/math/VectorUtils.java [hunk 13]
```diff
@@ -154,4 +158,38 @@ public static Vector copyPart(Vector v, int off, int len) {
...
+    }
+
+    /**
+     * Creates vector based on array of Doubles. If array contains null-elements then
+     * method returns sparse local on head vector. In other case method returns
+     * dense local on heap vector.
+     *
+     * @param values Values.
+     */
+    public static Vector of(Double[] values) {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/math/VectorUtils.java]: Why Double[]? Maybe double[]? or should add this kind of method?

File: modules/ml/src/main/java/org/apache/ignite/ml/math/VectorUtils.java [hunk 14]
```diff
@@ -154,4 +158,38 @@ public static Vector copyPart(Vector v, int off, int len) {
...
+        A.notNull(values, "values");
+
+        Vector answer = null;
+        if (Arrays.stream(values).anyMatch(Objects::isNull))
+            answer = new SparseLocalVector(values.length, StorageConstants.RANDOM_ACCESS_MODE);
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/math/VectorUtils.java]: maybe doing this more clearly with special flag (isSparse)? I don't catch the logic here
